calico/felix/selectors.py

```diff
@@ -0,0 +1,664 @@
+
+    def evaluate(self, labels):
+        try:
+            return labels[self.label_name]
+        except KeyError:
+            return NotPresent()
+
+    def __hash__(self):
+        return hash(self.label_name) * 37 + 0x5bce8abd
+
+    def __eq__(self, other):
+        return (type(other) == type(self) and
+                self.label_name == other.label_name)
+
+    def collect_str_fragments(self, fragment_list):
+        fragment_list.append(self.label_name)
+
+
+class HasOp(ExprNode):
+    """
+    AST node for a has(label) term.
+    """
+    __slots__ = ["label_name"]
+
+    def __init__(self, parse_str=None, location=None, tokens=None):
+        [self.label_name] = tokens
+
+    def evaluate(self, labels):
+        return self.label_name in labels
+
+    def __hash__(self):
+        return hash(self.label_name) * 37 + 0x742fe51e
+
+    def __eq__(self, other):
+        return (type(other) == type(self) and
+                self.label_name == other.label_name)
+
+    def collect_str_fragments(self, fragment_list):
+        fragment_list.append("has(%s)" % self.label_name)
+
+
+class Literal(ExprNode):
+    """
+    AST node for a string literal.
+    """
+    __slots__ = ["value"]
+
+    def __init__(self, parse_str=None, location=None, tokens=None):
+        [self.value] = tokens
+
+    def evaluate(self, labels):
+        return self.value
+
+    def __hash__(self):
+        return hash(self.__class__) * 37 + hash(self.value)
+
+    def __eq__(self, other):
+        return (type(other) == type(self) and
+                self.value == other.value)
+
+    def collect_str_fragments(self, fragment_list):
+        fragment_list.append(repr(self.value))
+
+
+class SetLiteral(ExprNode):
+    """
+    AST node for a set literal.
+    """
+    __slots__ = ["value"]
+
+    def __init__(self, parse_str=None, location=None, tokens=None):
+        self.value = frozenset(tokens)
+
+    def evaluate(self, labels):
+        return self.value
+
+    def __hash__(self):
+        return hash(self.__class__) * 37 + hash(self.value)
+
+    def __eq__(self, other):
+        return (type(other) == type(self) and
+                self.value == other.value)
+
+    def collect_str_fragments(self, fragment_list):
+        collect_set_string_fragments(fragment_list, self.value)
+
+
+def collect_set_string_fragments(fragment_list, the_set):
+    """
+    Appends string fragments to fragment_list to represent a set literal.
+    """
+    fragment_list.append("{")
+    first = True
+    for v in the_set:
+        if not first:
+            fragment_list.append(",")
+        else:
+            first = False
+        fragment_list.append(repr(v))
+    fragment_list.append("}")
+
+
+def equality_op(parse_str=None, location=None, tokens=None):
+    """Parse action for an '==' AST node."""
+    lhs, rhs = tokens
+    assert isinstance(lhs, Label) and isinstance(rhs, Literal)
+    return LabelToLiteralEquality(lhs.label_name, rhs.value)
```

Comment:
It wasn't clear to me why we seem to have changed paradigm here, from initialization in a class __init__ method, to initialization in a wrapper function.  From discussion it appears this is a bit historical - so with you to clean up as you think best.